File: 17xxx/17411.py
import bisect

MOD = 10**9 + 7

class LisSegmentTree:
    def __init__(self, N):
        self.tree = [(0, 0) for _ in range(4*N)]

    def add_size(self, a, b):
        if a[0] > b[0]:
            return a
        elif b[0] > a[0]:
            return b
        else:
            return (a[0], (a[1] + b[1]) % MOD)

    def query(self, start, end, left, right, i=1):
        if end < left or start > right:
            return (0, 0)
        
        if left <= start and end <= right:
            return self.tree[i]
        
        mid = (start + end) // 2
        return self.add_size(self.query(start, mid, left, right, i*2), self.query(mid+1, end, left, right, i*2+1))
    
    def update(self, start, end, idx, val, i=1):
        if start > idx or idx > end:
            return self.tree[i]
        
        if start == end:
            self.tree[i] = self.add_size(self.tree[i], val)
            return self.tree[i]
        
        mid = (start + end) // 2
        self.tree[i] = self.add_size(self.update(start, mid, idx, val, i*2), self.update(mid+1, end, idx, val, i*2+1))
        return self.tree[i]
    
    def getTree(self):
        return self.tree

import os, io, sys
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
print = sys.stdout.write # 줄바꿈 없음

N = int(input())
arr = list(map(int, input().split()))

unique = sorted(set(arr))
compressed = [bisect.bisect_left(unique, arr[i])+1 for i in range(N)]

segment_tree = LisSegmentTree(len(unique))
for i in range(N):
    val = compressed[i]
    length, count = segment_tree.query(1, len(unique), 1, val-1)
    if length == 0:
        count = 1
    segment_tree.update(1, len(unique), val, (length+1, count))

result = segment_tree.getTree()[1]
print(f"{result[0]} {result[1]}")

File: 17xxx/17411.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MOD = 1'000'000'007;

struct LisSegmentTree {
    int size;
    vector<pair<int, int>> tree;

    LisSegmentTree(int N) {
        size = 1;
        while (size < N)
            size <<= 1;
        tree.assign(2 * size, {0, 0});
    }

    pair<int, int> add_size(const pair<int, int> &a, const pair<int, int> &b) {
        if (a.first > b.first)
            return a;
        if (b.first > a.first)
            return b;
        return {a.first, (a.second + b.second) % MOD};
    }

    pair<int, int> query(int start, int end, int left, int right, int i = 1) {
        if (end < left || start > right) return {0, 0};
        if (left <= start && end <= right) return tree[i];
        int mid = (start + end) / 2;
        return add_size(query(start, mid, left, right, i * 2), query(mid + 1, end, left, right, i * 2 + 1));
    }

    pair<int, int> update(int start, int end, int idx, pair<int, int> val, int i = 1) {
        if (idx < start || idx > end) return tree[i];
        if (start == end) {
            tree[i] = add_size(tree[i], val);
            return tree[i];
        }
        int mid = (start + end) / 2;
        tree[i] = add_size(update(start, mid, idx, val, i * 2), update(mid + 1, end, idx, val, i * 2 + 1));
        return tree[i];
    }

    pair<int, int> getRoot() {
        return tree[1];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;
    vector<int> arr(N);
    for (int i = 0; i < N; i++) cin >> arr[i];

    vector<int> unique_arr = arr;
    sort(unique_arr.begin(), unique_arr.end());
    unique_arr.erase(unique(unique_arr.begin(), unique_arr.end()), unique_arr.end());
    vector<int> compressed;
    for (int i = 0; i < N; i++) {
        compressed.push_back((int)(lower_bound(unique_arr.begin(), unique_arr.end(), arr[i]) - unique_arr.begin()) + 1);
    }

    LisSegmentTree segment_tree(unique_arr.size());

    for (int i = 0; i < N; i++) {
        int val = compressed[i];
        pair<int, int> result = segment_tree.query(1, unique_arr.size(), 1, val - 1);
        if (result.first == 0) result.second = 1;
        segment_tree.update(1, unique_arr.size(), val, {result.first + 1, result.second});
    }

    pair<int, int> result = segment_tree.getRoot();
    cout << result.first << " " << result.second;

    return 0;
}
